### PR TITLE
RedisBloom T-Digest support

### PR DESCRIPTION
**Pull request by filipecosta90**
This PR adds a new sketching data-structure to RedisBloom -- The t-Digest. 

The t-Digest construction algorithm uses a variant of 1-dimensional k-means clustering to produce a very compact data structure that allows accurate estimation of quantiles. This t-Digest data structure can be used to estimate quantiles, compute other rank statistics, or even to estimate related measures like trimmed means.

- See [the original paper by Ted Dunning & Otmar Ertl](https://arxiv.org/abs/1902.04023) for more details on t-Digests.
- See the underlying [t-digest-c](https://github.com/RedisBloom/t-digest-c) implementation that we're using 
## What’s Inside

The following commands are implemented:
  - `TDIGEST.CREATE`: Allocate a new histogram
  - `TDIGEST.RESET`: Empty out a histogram and re-initialize it
  - `TDIGEST.ADD`: Add a value to the t-Digest with the specified count
  - `TDIGEST.MERGE`: Merge one t-Digest into another
  - `TDIGEST.CDF`:  Returns the fraction of all points added which are &le; x.
  - `TDIGEST.QUANTILE`: Returns an estimate of the cutoff such that a specified fraction of the data added to the t-Digest would be less than or equal to the cutoff.
  - `TDIGEST.MIN`: Get the minimum value from the histogram.  Will return __DBL_MAX__ if the histogram is empty
  - `TDIGEST.MAX`: Get the maximum value from the histogram.  Will return __DBL_MIN__ if the histogram is empty
  - `TDIGEST.INFO` : Returns compression, capacity, total merged and unmerged nodes, the total compressions 
made up to date on that key, and merged and unmerged weight.

## What’s done
- [x] document the added APIs
- [x] extended testing

### CODE REVIEW

**Review by itamarhaber — DISMISSED** [review]
LGTM (looks great to me) - did not review deeply though, just a cursory glance

**Review by ashtul — COMMENTED** [review]
Looks great!
A few minor suggestions.
This is a nice addition to RedisBloom.

File: src/rm_tdigest.c [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+ *
+ * Allocate the memory and initialize the t-digest.
+ *
+ * @param ctx Context in which Redis modules operate
+ * @param argv Redis command arguments, as an array of strings
+ * @param argc Redis command number of arguments
+ * @return REDISMODULE_OK on success, or REDISMODULE_ERR  if the command failed
+ */
+int TDigestSketch_Create(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    if (argc != 3) {
```
[the remark this answers is not in the document]
Comment (filipecosta90) [reply src/rm_tdigest.c:42]: @ashtul we should enforce the user to specify compression. It's a key parameter and the user should be aware of it. I've added further notes and a chart on the docs to fully explain the implication of compression vs accuracy.

File: src/rm_tdigest.c [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ | REDISMODULE_WRITE);
+
+    if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx, "ERR T-Digest: key does not exist");
+    } else if (RedisModule_ModuleTypeGetType(key) != TDigestSketchType) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+    }
```
[the remark this answers is not in the document]
Comment (filipecosta90) [reply src/rm_tdigest.c]: :+1: fixed in the latest commits

File: docs/TDigest_Commands.md [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+Allocate the memory and initialize the t-digest.
+
+```
+TDIGEST.CREATE {key} {compression}
+```
+
+#### Parameters:
+
+* **key**: The name of the sketch.
+* **compression**: The compression parameter. 100 is a common value for normal uses. 1000 is extremely large.
```
[the remark this answers is not in the document]
Comment (filipecosta90) [reply docs/TDigest_Commands.md:16]: @ashtul agree, added:
**Further notes on compression vs accuracy:**
Constructing a T-Digest requires a compression parameter which determines the size of the digest and accuracy of quantile estimation. 
The scaling of accuracy versus the compression parameter is illustrated in the following figure retrieved from ["Ted Dunning, The t-digest: Efficient estimates of distributions, Software Impacts,Volume 7,2021"](https://www.sciencedirect.com/science/article/pii/S2665963820300403).

Included image on docs.
PS: note that I'm against adding a default compression parameter given the user should be aware of it and master it.

[the remark this answers is not in the document]
Comment (ashtul) [reply docs/TDigest_Commands.md]: @filipecosta90 I completely agree about requiring the user to understand the compression parameter or at least to have to reach out to get advice

**Review by ashtul — APPROVED** [review]
Very nice work @filipecosta90 
Looking forward to watching how users would utilize it to solve real-world challenges.

**Comment by filipecosta90** [comment]
> Very nice work @filipecosta90
> Looking forward to watching how users would utilize it to solve real-world challenges.

Thank you @ashtul and @itamarhaber for the review :)

**Pull request merged**